modules/fundamental/src/purl/endpoints/mod.rs

```diff
@@ -33,9 +33,45 @@ pub fn configure(config: &mut utoipa_actix_web::service_config::ServiceConfig, d
         .service(base::all_base_purls)
         .service(version::get_versioned_purl)
         .service(get)
+        .service(get_multiple)
         .service(all);
 }
 
+async fn fetch_purl_details(
+    service: &PurlService,
+    db: &Database,
+    deprecated: trustify_module_ingestor::common::Deprecation,
+    identifiers: &[String],
+) -> Result<HashMap<String, PurlDetails>, String> {
+    let mut purls: Vec<Purl> = Vec::new();
+    let mut uuids: Vec<Uuid> = Vec::new();
+    for key in identifiers {
+        if key.starts_with("pkg:") {
+            let purl = Purl::from_str(key).map_err(|e| format!("Invalid purl '{}': {}", key, e))?;
+            purls.push(purl);
+        } else {
+            let id = Uuid::from_str(key).map_err(|e| format!("Invalid UUID '{}': {}", key, e))?;
+            uuids.push(id);
+        }
+    }
+    let mut result: HashMap<String, PurlDetails> = HashMap::new();
+    let purl_details = service
+        .purls_by_purl(purls, deprecated, db)
+        .await
+        .map_err(|e| format!("Failed to fetch purl details by purl: {}", e))?;
+    for detail in purl_details {
+        result.insert(detail.head.purl.to_string(), detail);
+    }
+    let uuid_details = service
+        .purls_by_uuid(uuids, deprecated, db)
+        .await
+        .map_err(|e| format!("Failed to fetch purl details by uuid: {}", e))?;
+    for detail in uuid_details {
+        result.insert(detail.head.uuid.to_string(), detail);
+    }
+    Ok(result)
+}
```

Comment:
I think this belongs in the service rather than the endpoint. And return an `Error` instead of a `String`, similar to how the other service functions do, e.g. `purls_by_purl`.

Style feedback: a for loop after a mutable collection is a smell to me. It looks like java code. :smile: There are almost always cleaner, safer alternatives in the `Iterator` trait, e.g. `fold`, `map`, `partition`, etc. It can sometimes be tricky to do in async code, but I think it's worth familiarizing yourself with `Iterator` to take advantage of it when you can.

Reply:
Guilty of being a Java developer 😅

Reply:
I have made the changes but I'm not sure if the way I'm creating the `uuids` and `purls` is more efficient as it has to iterate twice, given that I can't directly throw errors from closures.
Please, let me know if there's a better way to achieve something similar.

Reply:
Take a gander at the commit I just pushed, and let's discuss.